### ISSUE TITLE
Form validation: it allows only whitespaces in the required input and textarea form fields

### ISSUE THREAD

**Issue by sarasuri** [wet-boew/wet-boew#7557]
Hi,
It appears the jQuery form validation in the WET 4 current version (v4.0.21) allow only white spaces in the required form fields. When your enter only white spaces in a required field and validate the form the field with white spaces passes validation. 

I noticed in the previous versions of WET 4 (e.g. 4.0.20, 4.0,19,..) they didn't allow white spaces in a required form field and display an error message.

Shouldn't the current version validate the required form fields containing only white spaces and provide error message? Please confirm.

Version 4.0.20 screenshot

![form2](https://cloud.githubusercontent.com/assets/12256067/15506577/307a02fe-2196-11e6-9ce0-a742759ed941.gif)

Version 4.0.21 screenshot
![form1](https://cloud.githubusercontent.com/assets/12256067/15506596/37991548-2196-11e6-8586-429d6ded9f10.gif)

Thanks.

**Comment by hsrudnicki** [comment]
Does this remain still an issue, as well a help question or otherwise?  Please update the status of this again, thank you

**Comment by sarasuri** [comment]
Hi,
Yes, it is still an issue.  It is not a help question. I am seeking input on required input text fields accepting only white spaces are as valid data.
Shouldn't the form validation provide an error when users only enter white spaces in the required input text fields?

Thanks

**Comment by kpduncan** [comment]
Since I'm not part of WET here is a temp fix:

In additional-methods.js add (at end of page before `}));`):
```javascript
$.validator.addMethod( "minlengthNWS", function( value, element, param ) {
	return $.trim(value).length >= param;
}, $.validator.format( "Please enter at least {0} characters" ) );
```
In additional-methods.min.js add  (at end of page before `});`):
```javascript
a.validator.addMethod("minlengthNWS",function(b,c,d){return $.trim(b).length>=d},a.validator.format("Please enter at least {0} characters"))
```
How to use on form:

```HTML
<div class="wb-frmvld">
<form action="" method="post" novalidate="novalidate" >
<div class="form-group">
<label for="text" class="required"><span class="field-name">Text</span> <strong class="required">(required)</strong></label>
<input id="text" class="form-control" type="text" placeholder="Enter text" name="text" required="required" data-rule-minlengthNWS="1" data-rule-maxlength="250"/>
</div>
<div class="form-group">
<button class="btn btn-primary" type="submit" role="button" >Submit</button>
</div>
</form>
</div>
```

**Comment by kpduncan** [comment]
Can you add label 'Suggestion: Feature Request' @hsrudnicki thanks!

**Comment by hsrudnicki** [comment]
Updated as requested, and thank you for your status updates/cooperation

**Comment by sarasuri** [comment]
Thank you @kpduncan for the solution.

I noticed on the WET 4 Working example form validation page http://wet-boew.github.io/wet-boew/demos/formvalid/formvalid-en.html already using minlength rule for the required input fields. That is what I am using in my code as well.
For example:
 `<div class="wb-frmvld">
	<form action="#" method="get" id="validation-example">
		...
		<div class="form-group">
			<label for="fname1" class="required"><span class="field-name">First name</span> <strong class="required">(required)</strong></label>
			<input class="form-control" id="fname1" name="fname1" type="text" required="required" pattern=".{2,}" data-rule-minlength="2" />
		</div>`

I believe JS files jquery.validate.js and jquery.validate.min.js have the code to validate the minlength for the required fields.

If WET 4 group can modify above mentioned JS files wouldn't it be easy to modify the code in those files so that it won't allow only white spaces as valid data in the required fields. So that we don't need to change this attribute data-rule-minlength="2" in the HTML code.
Thanks.

**Comment by kpduncan** [comment]
Not a problem. The js files should be update to include my method or used to update this validation method `data-rule-minlength` all that is required is to add the `$.trim()` method.

**Comment by sarasuri** [comment]
Thank you @kpduncan.

I added the $.trim() method to the validation method data-rule-minlength in the JS files and it seems to work fine.

If those JS files can be updated with the solution in the future WET version it would be great.
Thanks.

**Comment by LaurentGoderre** [comment]
This is probably something that needs to be brought to jQuery validation

**Comment by kodecount** [comment]
I've open a issue to request the change [1876](https://github.com/jzaefferer/jquery-validation/issues/1876);

**Comment by LaurentGoderre** [comment]
I'm not sure it needs a new validator, perhaps an option in the way required works

**Comment by kodecount** [comment]
@LaurentGoderre You've lost me. I request I made askes to changed the core script to add $.trim() on minlength method.

**Comment by LaurentGoderre** [comment]
Are `minlength` and `required` the same?

**Comment by kodecount** [comment]
No. With `min-length` you can set a minimum amount of characters like for example this from the WET examples:

```html
<div class="form-group">
<label for="an1"><span class="field-name">Alphanumeric</span> (at least 4 characters) </label>
<input class="form-control" id="an1" name="an1" type="text" data-rule-alphanumeric="true" data-rule-minlength="4">
</div>
```

**Comment by LaurentGoderre** [comment]
The issues as posted was with the required rule though, not just minlength

**Comment by kodecount** [comment]
Yes it also accepts white space and is the same issue as with minlength look at [Line 1324](https://github.com/jzaefferer/jquery-validation/blob/master/src/core.js#L1324) I believe a quick $.trim would fix that too.

**Comment by kodecount** [comment]
@sarasuri @LaurentGoderre I believe this issue should be closed and a new issue open discussing wither WET wants to allow white space. jQuery Validation tools removed the `$.trim()` feature as some people were requesting to be allow to set fields with blank values like this `<input type="text" value="   " required/>`

**Comment by kodecount** [comment]
This was a possible solution provide by jQuery Validation group (but seem it maybe messy to mix with the way WET works):

```javascript
$("#myForm").validate({
  rules: {
    username: {
    required: true,
    // Using the normalizer to trim the value of the element
    // before validating it.
    normalizer: function(value) {
      // Trim the value of the input
      return $.trim(value);
    }
  }
});
```
